datasets/flwr_datasets/partitioner/semantic_partitioner.py

```diff
@@ -0,0 +1,493 @@
+    https://arxiv.org/abs/2110.14216
+
+    (Cited from section 4.1 in the paper)
+
+    Semantic partitioner's goal is to reverse-engineer the federated
+    dataset-generating process so that each client possesses semantically
+    similar data. For example, for the EMNIST dataset, we expect every client
+    (writer) to (i) write in a consistent style for each digit
+    (intra-client intra-label similarity) and (ii) use a consistent writing style
+    across all digits  (intra-client inter-label similarity). A simple approach
+    might be to cluster similar examples together  and sample client data from
+    clusters. However, if one directly clusters the entire dataset, the resulting
+    clusters may end up largely correlated to labels. To disentangle the effect
+    of label heterogeneity and semantic heterogeneity, we propose the following
+    algorithm to enforce intra-client intra-label similarity and intra-client
+    inter-label similarity in two separate stages.
+
+    • Stage 1: For each label, we embed examples using a pretrained neural
+    network (extracting semantic features), and fit a Gaussian Mixture Model
+    to cluster pretrained embeddings into groups. Note that this results
+    in multiple groups per label. This stage enforces intra-client
+    intra-label consistency.
+
+    • Stage 2: To package the clusters from different labels into clients,
+    we aim to compute an optimal multi-partite matching with cost-matrix
+    defined by KL-divergence between the Gaussian clusters. To reduce complexity,
+    we heuristically solve the optimal multi-partite matching by progressively
+    solving the optimal bipartite matching at each time for randomly-chosen
+    label pairs. This stage enforces intra-client inter-label consistency.
+
+    Parameters
+    ----------
+    num_partitions : int
+        The total number of partitions that the data will be divided into.
+    partition_by : str
+        Column name of the labels (targets) based on which Dirichlet sampling works.
+    efficient_net_type: int
+        The type of pretrained EfficientNet model.
+        Options: [0, 1, 2, 3, 4, 5, 6, 7], corresponding to EfficientNet B0-B7 models.
+    batch_size: int
+        The batch size for EfficientNet extracting embeddings.
+    pca_components: int
+        The number of PCA components for dimensionality reduction.
+    gmm_max_iter: int
+        The maximum number of iterations for the GMM algorithm.
+    gmm_init_params: str
+        The initialization method for the GMM algorithm.
+        Options: ["random", "kmeans", "k-means++"]
+    use_cuda: bool
+        Whether to use CUDA for computation acceleration.
+    shuffle: bool
+        Whether to randomize the order of samples. Shuffling applied after the
+        samples assignment to partitions.
+    seed: int
+        Seed used for dataset shuffling. It has no effect if `shuffle` is False.
```

Comment:
```suggestion
        Seed used for dataset shuffling and as random_state in PCA and GMM. 
        It has no effect on shuffling if `shuffle` is False. It is always passed to 
        the models mentioned before.
```

Reply:
Alternatively, we can have two additional parameters specifying the seed values for the PCA and GMM.

Reply:
Agree. I will add corresponding test case too.